tests/data_context/test_data_context_config_variables.py

```diff
@@ -481,11 +483,357 @@ def test_escape_all_config_variables(empty_data_context_with_config_variables):
+            value=multi_value_str, skip_if_substitution_variable=False
+        )
+        == escaped_multi_value_str
+    )
+
+    multi_value_str2 = "$USER:pas$word@$HOST:${PORT}/${DATABASE}"
+    escaped_multi_value_str2 = r"\$USER:pas\$word@\$HOST:\${PORT}/\${DATABASE}"
+    assert (
+        context.escape_all_config_variables(
+            value=multi_value_str2, skip_if_substitution_variable=True
+        )
+        == multi_value_str2
+    )
+    assert (
+        context.escape_all_config_variables(
+            value=multi_value_str2, skip_if_substitution_variable=False
+        )
+        == escaped_multi_value_str2
+    )
+
+    multi_value_str3 = "USER:pas$word@$HOST:${PORT}/${DATABASE}"
+    escaped_multi_value_str3 = r"USER:pas\$word@\$HOST:\${PORT}/\${DATABASE}"
+    assert (
+        context.escape_all_config_variables(
+            value=multi_value_str3, skip_if_substitution_variable=True
+        )
+        == escaped_multi_value_str3
+    )
+    assert (
+        context.escape_all_config_variables(
+            value=multi_value_str3, skip_if_substitution_variable=False
+        )
+        == escaped_multi_value_str3
+    )
+
+    # dict
+    value_dict = {
+        "drivername": "postgresql",
+        "host": "${HOST}",
+        "port": "5432",
+        "username": "postgres",
+        "password": "pass$word1",
+        "database": "$postgres",
+        "sub_dict": {
+            "test_val_no_escaping": "test_val",
+            "test_val_escaping": "te$t_val",
+            "test_val_substitution": "$test_val",
+            "test_val_substitution_braces": "${test_val}",
+        },
+    }
+    escaped_value_dict = {
+        "drivername": "postgresql",
+        "host": r"\${HOST}",
+        "port": "5432",
+        "username": "postgres",
+        "password": r"pass\$word1",
+        "database": r"\$postgres",
+        "sub_dict": {
+            "test_val_no_escaping": "test_val",
+            "test_val_escaping": r"te\$t_val",
+            "test_val_substitution": r"\$test_val",
+            "test_val_substitution_braces": r"\${test_val}",
+        },
+    }
+    escaped_value_dict_skip_substitution_variables = {
+        "drivername": "postgresql",
+        "host": "${HOST}",
+        "port": "5432",
+        "username": "postgres",
+        "password": r"pass\$word1",
+        "database": "$postgres",
+        "sub_dict": {
+            "test_val_no_escaping": "test_val",
+            "test_val_escaping": r"te\$t_val",
+            "test_val_substitution": "$test_val",
+            "test_val_substitution_braces": "${test_val}",
+        },
+    }
+    assert (
+        context.escape_all_config_variables(
+            value=value_dict, skip_if_substitution_variable=False
+        )
+        == escaped_value_dict
+    )
+    assert (
+        context.escape_all_config_variables(
+            value=value_dict, skip_if_substitution_variable=True
+        )
+        == escaped_value_dict_skip_substitution_variables
+    )
```

Comment:
Love the thoroughness here! I was just looking for these test cases, and here they are.